src/mcmc_gibbs.cpp

```diff
@@ -0,0 +1,138 @@
+                val += strength * (psi);
+            }
+        }
+    }
+    return val;
+}
+
+/*
+ * Add specific constraint weights & return the cumulative weight vector
+ */
+double calc_gibbs_tgt(const subview_col<uword> &plan, int n_distr, int V,
+                      IntegerVector districts, NumericVector &psi_vec, const uvec &pop,
+                      double parity, const Graph &g, List constraints) {
+    if (constraints.size() == 0) return 0.0;
+    double log_tgt = 0;
+
+    log_tgt += add_constraint("pop_dev", constraints, districts, psi_vec,
+                              [&] (List l, int distr) -> double {
+                                  return eval_pop_dev(plan, distr,
+                                                      pop, parity);
+                              });
+
+    log_tgt += add_constraint("splits", constraints, districts, psi_vec,
+                              [&] (List l, int distr) -> double {
+                                  return eval_splits(plan, distr, as<uvec>(l["admin"]), l["n"]);
+                              });
+
+    log_tgt += add_constraint("multisplits", constraints, districts, psi_vec,
+                              [&] (List l, int distr) -> double {
+                                  return eval_multisplits(plan, distr, as<uvec>(l["admin"]), l["n"]);
+                              });
+
+    log_tgt += add_constraint("segregation", constraints, districts, psi_vec,
+                              [&] (List l, int distr) -> double {
+                                  return eval_segregation(plan, distr, as<uvec>(l["group_pop"]), as<uvec>(l["total_pop"]));
+                              });
+
+    log_tgt += add_constraint("grp_pow", constraints, districts, psi_vec,
+                              [&] (List l, int distr) -> double {
+                                  return eval_grp_pow(plan, distr, as<uvec>(l["group_pop"]),
+                                                      as<uvec>(l["total_pop"]), as<double>(l["tgt_group"]),
+                                                      as<double>(l["tgt_other"]), as<double>(l["pow"]));
+                              });
+
+    log_tgt += add_constraint("grp_hinge", constraints, districts, psi_vec,
+                              [&] (List l, int distr) -> double {
+                                  return eval_grp_hinge(plan, distr, as<vec>(l["tgts_group"]),
+                                                        as<uvec>(l["group_pop"]), as<uvec>(l["total_pop"]));
+                              });
+
+    log_tgt += add_constraint("grp_inv_hinge", constraints, districts, psi_vec,
+                              [&] (List l, int distr) -> double {
+                                  return eval_grp_inv_hinge(plan, distr, as<vec>(l["tgts_group"]),
+                                                            as<uvec>(l["group_pop"]), as<uvec>(l["total_pop"]));
+                              });
```

Comment:
see above. should be basically
```cpp
    log_tgt += add_constraint("grp_inv_hinge", constraints, districts, psi_vec,
                              [&] (List l, int distr) -> double {
                                  return eval_grp_hinge(plan, distr, 1.0 - as<vec>(l["tgts_group"]),
                                                        as<uvec>(l["total_pop"]) - as<uvec>(l["group_pop"]), 
                                                        as<uvec>(l["total_pop"]));
                              });
```
though might be good to precompute these on the R side when the constraint list is being set up. 

come to think of it, the `inverse=TRUE` flag to `add_constr_grp_hinge()` would take care of this by just inverting if the flag is true & then requiring no c++ changes or new code to maintain

Reply:
Ah, now I see what you meant. Happy to adjust this, though I think on the user side, it is easier to have two separate function. Can use that to do the precomputation on the R. Either way, the precomputation makes sense to avoid additional compiling and C++

Reply:
yeah I'm ok w/o the `inverse=TRUE` but basically the R function should just precompute the opposite and then the c++ should call the same `eval_grp_hinge` for both

Reply:
Okay, that's updated now